lambda/lambda.mli

```diff
@@ -179,6 +179,9 @@ type primitive =
   (* Array operations *)
   | Pmakearray of array_kind * mutable_flag * locality_mode
   | Pmakearray_dynamic of array_kind * locality_mode
+  (** For [Pmakearray_dynamic], if the array kind specifies an unboxed
+      product, the float array optimization will never apply -- even if such
+      unboxed product is a singleton. *)
```

Comment:
Actually, this case might be ruled out by the frontend.  Let's discuss with @ccasin.

Reply:
This is ruled out by the front-end (and the array specialization code in typeopt).  Note the comment on `array_kind` in this file documenting the invariant that product array kinds have at least two elements.

It's still a little weird that `#(int32# * int32#) array` isn't packed tightly like `int32# array`, but I'm content to wait for other work on nesting before attempting that.

Reply:
We can probably deal with the packing at the same time as allowing unboxed vec128s to be put in product arrays.  In both cases I think the indexing code needs adjusting, but this isn't completely straightforward.

Reply:
In any case I think it would be even more surprising for `#(int32# * int32#)` to have different packing than `{ a : int32#; b : int32# }`, so we should do all of the packing at once.